core/lib/contract_verifier/src/etherscan/mod.rs

```diff
@@ -0,0 +1,459 @@
+            .await?;
+
+        Ok(Some(request_id))
+    }
+
+    async fn await_verification(
+        &mut self,
+        verification_request: &VerificationRequest,
+        verification_id: String,
+    ) -> Result<(), VerifierError> {
+        let mut get_status_attempts = 0;
+        loop {
+            get_status_attempts += 1;
+            tracing::info!(
+                "Fetching contract verification status, address = {:#?}, attempt = {}",
+                verification_request.req.contract_address,
+                get_status_attempts
+            );
+            let result = self.client.get_verification_status(&verification_id).await;
+            match result {
+                Ok(_) => break,
+                Err(EtherscanError::VerificationPending) => {
+                    if tokio::time::timeout(
+                        POLL_VERIFICATION_RESULT_INTERVAL,
+                        self.stop_receiver.changed(),
+                    )
+                    .await
+                    .is_ok()
+                    {
+                        return Err(VerifierError::Canceled);
+                    }
+                    if get_status_attempts > MAX_POLL_VERIFICATION_RESULT_ATTEMPTS {
+                        tracing::warn!(
+                            "Get verification status timed out for contract with address {:#?}",
+                            verification_request.req.contract_address
+                        );
+                        return Err(ProcessingError::VerificationStatusPollingTimeout.into());
+                    }
+                    continue;
+                }
+                Err(e) => return Err(e.into()),
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn verify(
+        &mut self,
+        verification_request: VerificationRequest,
+        etherscan_verification_id: Option<String>,
+    ) -> Result<(), VerifierError> {
+        let req = verification_request.req.clone();
+        let is_verified = self
+            .client
+            .is_contract_verified(req.contract_address)
+            .await?;
+
+        if is_verified {
+            tracing::info!(
+                "Contract with address {:#?} is already verified",
+                req.contract_address
+            );
+            return Ok(());
+        }
+        let verification_id = match etherscan_verification_id {
+            Some(id) => Some(id),
+            None => {
+                let request_id = self
+                    .send_verification_request(&verification_request)
+                    .await?;
+                // Wait for the verification to be processed
+                if tokio::time::timeout(
+                    POLL_VERIFICATION_RESULT_INTERVAL,
+                    self.stop_receiver.changed(),
+                )
+                .await
+                .is_ok()
+                {
+                    return Err(VerifierError::Canceled);
+                }
```

Comment:
Do we need this sleep? Wouldn't it be handled by `await_verification`? Looks a bit out of place.

Reply:
There are two scenarios in which we get the `verification_id`:

1. It's already in the request (this means the verification was sent before, and the execution got interrupted). In this case, we can poll for results right away.
2. It's not in the request, and we've just received it from the verification request. In this case, we need to pause, as it takes time for Etherscan to process the request. We can call without pausing, but the verification will still be pending. To avoid spamming Etherscan, I wait for some time.

We can let `await_verification` know whether to wait before the first check. Do you think that will make it clearer? To me, having a pause right after the verification request makes sense, as I see them together and understand the reasoning.

Reply:
I think the case "verification request was sent before, and the execution got interrupted" can be considered rare enough, and we can just add a wait before each check iteration. It would also make the logic for sleeping in `await_verification` simpler (we won't have to do `if attempt < MAX_ATTEMPTS`. WDYT?